sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala

```diff
@@ -258,21 +275,29 @@ class SortBasedAggregator(
         if (hasNextInput || hasNextAggBuffer) {
           // Find smaller key of the initialAggBufferIterator and initialAggBufferIterator
           groupingKey = findGroupingKey()
-          result = new AggregationBufferEntry(groupingKey, makeEmptyAggregationBuffer)
+          updateResult = new AggregationBufferEntry(
+            groupingKey, makeEmptyAggregationBufferForSortBasedUpdateAggFunctions)
+          finalResult = new AggregationBufferEntry(
+            groupingKey, makeEmptyAggregationBufferForSortBasedMergeAggFunctions)
 
           // Firstly, update the aggregation buffer with input rows.
           while (hasNextInput &&
             groupingKeyOrdering.compare(inputIterator.getKey, groupingKey) == 0) {
-            processRow(result.aggregationBuffer, inputIterator.getValue)
+            processRow(updateResult.aggregationBuffer, inputIterator.getValue)
             hasNextInput = inputIterator.next()
           }
 
+          // This step ensures that the contents of the updateResult aggregation buffer are
+          // merged with the finalResult aggregation buffer to maintain consistency
+          serializeBuffer(updateResult.aggregationBuffer)
+          mergeAggregationBuffers(finalResult.aggregationBuffer, updateResult.aggregationBuffer)
           // Secondly, merge the aggregation buffer with existing aggregation buffers.
           // NOTE: the ordering of these two while-block matter, mergeAggregationBuffer() should
           // be called after calling processRow.
           while (hasNextAggBuffer &&
             groupingKeyOrdering.compare(initialAggBufferIterator.getKey, groupingKey) == 0) {
```

Comment:
I think the data flow is
1. start with hash agg, create and initialize an empty buffer for each group key, and consume input records by calling `update`
2. fallback to sort based agg
3. create and initialize an empty buffer for each group key, and consume input records by calling `update`.
4. when all the input records for one group key are consumed up, call `merge` to merge the agg buffer that is generated by hash agg.

This data flow is fine most of the time, but is problematic for hive UDAF, because it doesn't support INIT -> UPDATE -> MERGE -> FINISH.

That said, this patch may cause perf regression for normal UDAFs. We should only do it for hive UDAF, to switch from `INIT -> UPDATE -> MERGE -> FINISH` to `INIT -> UPDATE -> FINISH, INIT -> MERGE -> FINISH`

Reply:
Hi @cloud-fan, what you said here is really killing! With Spark 2.4.1 by turning this conf off I see the DataSketches hll issue is solved: `--conf spark.sql.execution.useObjectHashAggregateExec=false`. Which basically disable the hash agg attempt. But as you said, this is downgrading the general performance (not so bad when people have a sense about how big their data is).
However my question is, is it possible that it can be automatically done for the framework to recognize Hive UDAF and only apply its way to it? e.g. For such a query `select year, count(month), hive_udaf_count(month) from ... group by year` I want the Spark count() to behave as the way INIT -> UPDATE -> MERGE -> FINISH, while the hive_udaf_count() to behave as the way INIT -> UPDATE -> FINISH, INIT -> MERGE -> FINISH